proposals/2024-05-23_controller_engine_v2.md

```diff
@@ -0,0 +1,304 @@
+    important now that mapping authors are starting to reuse functionality
+    across different hardware devices (see Numark NS6II and Mixtrack variants or
+    similarity between different pioneer controllers)
+ 3. Current mappings modify global state and assume that they are the only
+    mapping running. This needs to be fixed in order to support multiple
+    mappings running at the same time in the same "execution engine".
+ 4. Device hotplug is currently not possible / hard to implement. This is
+    primarily because of the underlying protocol IO layer not willing to support
+    it. This document proposes an alternative approach that lets us more easily
+    switch the underlying library.
+ 5. Allow changing the manifest format. Many people have expressed distaste in
+    the XML manifest format. Careful implementation of the new format should
+    allow experimentation with different formats.
+ 6. Changing out infrastructure like this is virtually impossible. The proposed
+    architecture is designed to be modular and allows for all the required
+    components to be implemented gradually.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Allow multiple controllers to be used at the same time.
+* Mappings should be able to organize themselves into modules.
+* IO protocols should easily be changeable and not not be tied to the mapping.
+* Allow easy, gradual implementation of the system.
+
+### Audience
+
+This is primarily targeted at developers and power users who are interested in
+the controller engine. End-users only interested in the point-and-click mapping
+editor should not experience any significant changes.
+
+## Non-Goals
+
+* Replace `ControllerScriptEngine` in the short term until we are confident in
+  the new design.
+* make changes to the scratching code as it is orthogonal to this proposal.
+* Make changes to the effects system as it is orthogonal to this proposal.
+
+## How
+
+Explain the full overview of the proposed solution. Some guidelines:
+
+### File Structure
+
+Built-in mappings reside in `res/controllers/` where each mapping is a directory
+containing at least a manifest file named `manifest.xxx` (where `.xxx` is the
+language-specific file extension (eg. `.xml`, `.yml`, `.json`, etc)).
+Additionally, `res/controllers/lib` contains shared modules that can be used by
+multiple mappings (such as componentsJS).
+
+### Manifest
+
+The manifest is a declarative file that describes the mapping. It contains the
+following information:
+
+* Sources and Sinks
+* Execution Engines
+  * Used Mixxx APIs
+* Capabilities
+
+### Sources and Sinks
+
+Sources and Sinks are the endpoints of the controller. They are defined in the
+manifest and are used to define the IO protocol of the controller. They are
+characterized by their protocol (eg. MIDI, HID, Bulk, etc), their direction (eg.
+input, output, bidirectional) and optionally explicitly their underlying
```

Comment:
Can you elaborate? 
These sections should allow the mapping to declare things like "I need an HID endpoint and I intend to read from it". Then if no hardware with that capability is found, the mapping can not be loaded in the first place.

Reply:
Every USB device has "Endpoint Descriptors" ( https://beyondlogic.org/usbnutshell/usb5.shtml ), which contain this information per enpoint. If I know the supported device type, why should I store a copy of this information in the mapping files?

Reply:
good point. Primarily so the mapping can declare stable identifiers to address them. What do you propose instead?

Reply:
I propose to assign the device type to mapping by "Manufacuter ID" and "Product ID" and read out all information of the device, which we can gather via userspace APIs. Only what is not accessible on all supported systems, shall be stored in the mapping files.

Reply:
I think there is a risk of doing a 1-to-1 relationship here. Having the IO driver respecting the USB descriptor is definitely a must have, but it would make sense to also explicately define the direction the mapping is interested with. This would help us planning for vendor which may be using not so compliant descriptor (BULK being probably the most at risk here, since it already involve a custom API made by the vendor), meaning that the port could be used for more than just receiving/sending mapping relevant data.

Reply:
I'm still not convinced how this direction property should be clearly seperated, e.g.:
- A BULK mapping might need to send a memory address before reading a value. While it is only done for reading, it technically would send a BULK message with an address.
- HID allows to read the value of the OutputReport back from the device (e.g. for seamless initialization at hot-plugging) Does this make any OutputReport to a bidirectional Source/Sink?
For what is this information needed for?

Why do we need to enforce this by the top level architecture. I fear that this will restrict us in the future, as we will always have to force all USB protocols into this arbitrary structure.

Reply:
I think your concerns are valid. So the consequence is that we only apply the source/sink abstraction to midi. So no abstraction over these protocols. That may be fine, but then we would have to reimplement the PnC logic for each protocol. Do you have some other abstraction in mind?

Reply:
We need an entity to assign to the execution engine here, but this must be an abstract definition to be future proof. It should cover cases like:
- An USB interface for BULK mode
- MIDI1 and MIDI2 devices
- An USB interface, that appears as two MIDI "devices" with  https://github.com/mixxxdj/mixxx/blob/a65bc0bf3ac13114b2d7a8a2474394768a321c27/src/controllers/midi/portmidienumerator.cpp#L76C1-L93C1
- An USB HID interface with two top-level-collections, which appears under macOS and Linux as one HIDAPI device
- An USB HID interface with two top-level-collections, which appears under Windows as two HIDAPI devices
- A MIDI device that is accessible over two physical interfaces (like the DDJ-FLX4 with MIDI over BluetoothLE and USB)

Reply:
Thanks for the comprehensive collection, but it doesn't answer my question. What do you imagine the common API between these to look like? We need some abstraction unless we want to reimplement common features such as the PnC editor for each protocol.